ocaml/xapi/xapi_pool_update.ml

```diff
@@ -532,21 +532,50 @@ let resync_host ~__context ~host =
     |> List.iter (fun self -> destroy ~__context ~self)
   end
 
-let path_from_uri uri =
+let path_and_host_from_uri uri =
   (* remove any dodgy use of "." or ".." NB we don't prevent the use of symlinks *)
-  String.sub_to_end uri (String.length Constants.get_pool_update_download_uri)
-  |> Filename.concat !Xapi_globs.host_update_dir
-  |> Uri.pct_decode
-  |> Stdext.Unixext.resolve_dot_and_dotdot
+  let host_and_path = String.sub_to_end uri (String.length Constants.get_pool_update_download_uri) in
+  match String.split ~limit:2 '/' host_and_path with
+  | [host; untrusted_path] ->
+    let resolved_path = untrusted_path
+      |> Filename.concat !Xapi_globs.host_update_dir
+      |> Uri.pct_decode
+      |> Stdext.Unixext.resolve_dot_and_dotdot
+    in
+    (host,resolved_path)
+  | _ -> failwith "Expecting both host and path in the uri"
+
+let proxy_request req s host_uuid =
+  Server_helpers.exec_with_new_task "pool_update_proxy_request" (fun __context ->
+    let host =
+      try Some (Db.Host.get_by_uuid ~__context ~uuid:host_uuid) with _ -> None
```

Comment:
I don't know enough about xapi here, so this is probably not an issue at all, but is `get_by_uuid` safe to untrusted data, i.e. is there any possibility an attacker could encode something in it that could lead to a SQL injection (or equivalent, as I know we aren't actually using a SQL DB)?

Reply:
An `xe host*` API would also at some point do a get_by_uuid, all it does is a lookup in a map/list matching the uuid exactly as it is, so this API is safe against injection.
There is another API get_all_records_where that allows specifying a filter string, but even if you manage to inject it that expression is interpreted by XAPI and all you can do is match on fields and perform boolean operations, it is never given to a C API call, or some DB.